flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CoordinatorEventsToStreamOperatorRecipientExactlyOnceITCase.java

```diff
@@ -0,0 +1,385 @@
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.api.operators.StreamOperatorParameters;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.SerializedValue;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Integration test case that validates the exactly-once mechanism for operator events sent around
+ * checkpoint. This class is an extension to {@link CoordinatorEventsExactlyOnceITCase}, further
+ * verifying the exactly-once semantics of events in the following conditions:
+ *
+ * <h2>Stream operator recipient</h2>
+ *
+ * <p>In {@link CoordinatorEventsExactlyOnceITCase}, the test cases focus on verifying the
+ * correctness of operator coordinator's behavior. It uses a custom {@link AbstractInvokable}
+ * subclass that mocks the behavior of coordinator's recipients. This test class uses actual stream
+ * operators as the recipient of coordinator events, verifying that stream operators can correctly
+ * handle received operator events and inform coordinators of completing checkpoint.
+ *
+ * <h2>Non-source stream task</h2>
+ *
+ * <p>In {@link CoordinatorEventsExactlyOnceITCase}, the two tested tasks are executed independently
+ * of each other. They do not have the upstream-downstream relationship as operators usually do in a
+ * streaming job, and thus both of them are treated as source tasks. This test class further
+ * verifies situations when the tested operators are not sources, which means when checkpoint
+ * barriers are injected into sources, these operators may not have started checkpoint yet.
+ *
+ * <h2>Unaligned checkpoint</h2>
+ *
+ * <p>This class tests both aligned and unaligned checkpoints to verify that the correctness of the
+ * event delivery behavior around checkpoint is not affected by this condition.
+ *
+ * <h2>Non-global failure</h2>
+ *
+ * <p>In {@link CoordinatorEventsExactlyOnceITCase}, failures occur at the coordinators' side, so
+ * they will cause the whole Flink job to fail over. In this class, test cases are added when there
+ * might only be fail-overs on the subtasks' side, while the coordinators are not affected. In this
+ * case the production infrastructure code needs to work together with user code (implementations
+ * inside the coordinator and operator subclass) to ensure the exactly-once semantics of operator
+ * event delivery.
+ */
+public class CoordinatorEventsToStreamOperatorRecipientExactlyOnceITCase
+        extends CoordinatorEventsExactlyOnceITCase {
+    private static final AtomicBoolean shouldCloseSource = new AtomicBoolean(false);
+
+    private static final int numEvents = 100;
```

Comment:
Is this variable name consistent with other `private static final` variable names in the repo?